homeassistant/components/samsungtv/bridge.py

```diff
@@ -409,70 +418,139 @@ def _send_key(self, key: str, key_type: str | None = None) -> None:
-                LOGGER.debug("WebSocketException, OSError %s", err.__repr__())
-                self._remote = None
+            except (WebSocketException, AsyncioTimeoutError, OSError) as err:
+                LOGGER.debug("WebSocketException, OSError %s", err)
             else:
-                LOGGER.debug(
-                    "Created SamsungTVWSBridge for %s (%s)", CONF_NAME, self.host
-                )
-                if self.token != self._remote.token:
+                LOGGER.info("Created SamsungTVWSBridge for %s", self.host)
+                self._remote = remote
+                if self.token != remote.token:
                     LOGGER.debug(
                         "SamsungTVWSBridge has provided a new token %s",
-                        self._remote.token,
+                        remote.token,
                     )
-                    self.token = self._remote.token
+                    self.token = remote.token
                     self._notify_new_token_callback()
         return self._remote
 
-    async def async_close_remote(self) -> None:
-        """Close remote object."""
-        await self.hass.async_add_executor_job(self._close_remote)
+    async def _websocket_event(self, event: str, response: Any) -> None:
+        """Handle websocket event."""
+        LOGGER.debug("SamsungTVWS websocket event: %s", response)
+        if event == "ed.installedApp.get":
+            while self._app_list_futures:
+                self._app_list_futures.pop().set_result(response)
 
-    def _close_remote(self) -> None:
+    async def async_close_remote(self) -> None:
         """Close remote object."""
         try:
             if self._remote is not None:
                 # Close the current remote connection
-                self._remote.close()
+                await self._remote.close()
             self._remote = None
         except OSError:
             LOGGER.debug("Could not establish connection")
+
+
+# This will be moved to the external library
+# It is here to allow testing
+class SamsungTVWSAsyncConnection(BaseSamsungTVWSAsyncConnection):  # type: ignore[misc]
+    """Custom class to extend SamsungTVWSAsyncConnection."""
+
+    connection: WebSocketClientProtocol | None
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Initialise class."""
+        super().__init__(*args, **kwargs)
+        self._recv_loop: Task | None = None
+
+    async def open(self) -> WebSocketClientProtocol:
+        """Open websocket."""
+        if self.connection:
+            # someone else already created a new connection
+            return self.connection
+
+        self.connection = await super().open()
+        return self.connection  # type: ignore[return-value]
+
+    async def start_listening(
+        self, callback: Callable[[str, Any], Awaitable[None]]
+    ) -> None:
+        """Open, and start listening."""
+        if self.connection is None:
+            self.connection = await self.open()
```

Comment:
```suggestion
        if self.connection is None:
            await self.open()
```

Set it in one place as it makes it harder to check for race safety otherwise